File: python/src/application.py
import inspect
import json
import os
from typing import Any


class Application:
    """Reads a JSON object file, sorts its top-level keys (and any nested
    dict/list values) ascending or descending, and rewrites it as pretty
    JSON."""

    class InvalidFilenameError(Exception):
        pass

    class InvalidOrderError(Exception):
        pass

    @classmethod
    def run(cls, dirname: str = '', filename: str = '',
            order: Any = 'asc') -> None:
        instance = cls(dirname=dirname, filename=filename, order=order)
        instance.validate_filename()
        instance.validate_order()
        instance._run()

    def __init__(self, dirname: str = '', filename: str = '',
                 order: Any = 'asc') -> None:
        self._dirname = dirname
        self._filename = filename
        self._filepath = os.path.join(dirname, filename)
        self._order = order

    def validate_filename(self) -> str:
        if not self._filename:
            raise self.InvalidFilenameError('Filename must be provided.')
        return self._filename

    def validate_order(self) -> str:
        if not isinstance(self._order, str) or \
                self._order not in ('asc', 'desc'):
            raise self.InvalidOrderError(
                'Order option must be either asc or desc.')
        return self._order

    # private

    def _run(self) -> None:
        self._output(f'Start exporting JSON data in {self._filepath}')
        os.makedirs(self._dirname, exist_ok=True)
        if not os.path.isfile(self._filepath):
            with open(self._filepath, 'w') as f:
                f.write('')
        dumped = self._dump_converted_json_data_with_sorting()
        with open(self._filepath, 'w') as f:
            f.write(dumped)
        self._output(f'Done export JSON data in {self._filepath} 🎉')

    def _json_data(self) -> dict[str, Any]:
        with open(self._filepath) as f:
            return json.load(f)

    def _converted_json_data_with_sorting(self) -> dict[str, Any]:
        return dict(sorted(
            self._json_data().items(),
            reverse=(self._order == 'desc'),
        ))

    def _dump_converted_json_data_with_sorting(self) -> str:
        sorted_hash = {
            key: self._sort_value(value)
            for key, value in
            self._converted_json_data_with_sorting().items()
        }
        return json.dumps(sorted_hash, ensure_ascii=False, indent=2)

    def _sort_value(self, value: Any) -> Any:
        if isinstance(value, dict):
            return dict(sorted(
                value.items(), reverse=(self._order == 'desc')))
        if isinstance(value, list):
            return sorted(value, reverse=(self._order == 'desc'))
        return value

    def _test_env(self) -> bool:
        stack = inspect.stack()
        if not stack:
            return False
        return 'pytest' in os.path.basename(stack[-1].filename)

    def _output(self, message: str) -> None:
        if not self._test_env():
            print(message)

File: ruby/src/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'json'

class Application
  class InvalidFilenameError < StandardError; end
  class InvalidOrderError < StandardError; end

  # @rbs dirname: String
  # @rbs filename: String
  # @rbs order: untyped
  # @rbs return: void
  def self.run(dirname: '', filename: '', order: :asc)
    order    = order.to_sym if order.respond_to?(:to_sym)
    instance = new(dirname:, filename:, order:)
    instance.validate_filename!
    instance.validate_order!
    instance.run
  end

  # @rbs dirname: String
  # @rbs filename: String
  # @rbs order: untyped
  # ⚠️ `order` expects a class which can be converted to Symbol by `to_sym` method(e.g. String, Symbol).
  # ⚠️ Otherwise, it will raise an error when validating order.
  # @rbs return: void
  def initialize(dirname: '', filename: '', order: :asc)
    @dirname  = dirname
    @filename = filename
    @filepath = File.join(dirname, filename)
    @order    = order
  end

  # @rbs return: String
  def validate_filename!
    raise InvalidFilenameError, 'Filename must be provided.' if filename.empty?

    filename
  end

  # @rbs return: Symbol
  def validate_order!
    case order
    when :asc, :desc
      order
    else
      raise InvalidOrderError, 'Order option must be either :asc or :desc'
    end
  end

  # @rbs return: void
  def run
    output "Start exporting JSON data in #{filepath}"
    FileUtils.mkdir_p(dirname)
    FileUtils.touch(filepath) unless File.exist?(filepath)
    File.write(filepath, dump_converted_json_data_with_sorting)
    output "Done export JSON data in #{filepath} 🎉"
  end

  private

  attr_reader :dirname, :filename, :filepath, :order

  # @rbs return: Hash[String, untyped]?
  def json_data
    File.open(filepath) { |f| JSON.parse(f) }
  end

  # @rbs return: Hash[String, untyped]?
  def converted_json_data_with_sorting
    sorted_json_data = json_data&.sort
    order == :asc ? sorted_json_data&.to_h : sorted_json_data&.reverse&.to_h
  end

  # @rbs hash: Hash[String, untyped]
  # @rbs return: String
  def dump_converted_json_data_with_sorting(hash = {})
    sorted_hash = converted_json_data_with_sorting&.each_with_object(hash) do |(key, value), acc|
      acc[key] = sort_value(value)
    end
    JSON.pretty_generate(sorted_hash)
  end

  # @rbs value: untyped
  # @rbs return: untyped
  def sort_value(value)
    case value
    when Hash
      order == :asc ? value.sort.to_h : value.sort.reverse.to_h
    when Array
      order == :asc ? value.sort : value.sort.reverse
    else
      value
    end
  end

  # @rbs return: bool
  def test_env?
    caller(0..0).first.split('/').last.include?('minitest.rb')
  end

  # @rbs message: String
  # @rbs return: void
  def output(message)
    puts message unless test_env?
  end
end
